### PR TITLE
feat: thorchain repayments display approval or gas fee

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

See https://github.com/shapeshift/web/pull/7088#pullrequestreview-2114666233 for context:

> We still have the issue of the network fee skeleton on approval (i.e. not showing the approval fee estimate to the user), but that is not expected for this PR (though as you noted, we should absolutely do in a follow up).

This does precisely that, while doing some react-query cleanup following recent learnings. 

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

- continuation of the closed https://github.com/shapeshift/web/issues/7083

> What protocols, transaction types or contract interactions might be affected by this PR?

Medium - theoretically low, but inherent medium risks in the change of logic - if we got EVM/token checks wrong somewhere, this could rug lending repayments

## Testing

Note, there is no need to broadcast a Tx for this, though absolutely fine to do for paranoia on a cheap EVM chain

- When approval is required for a token repayment, "Approval Fee" is displayed instead of "Gas Fee"
- The fee row gracefully switches to "Gas Fee" after allowance is granted, and loading states are handled gracefully
- Non-token repayment quotes show no regressions

### Engineering

- ^

### Operations

- ^

## Screenshots (if applicable)

https://jam.dev/c/a7de2dfa-b436-4ffa-990c-8df9cf095215

### CODE REVIEW

File: src/hooks/mutations/useApprove.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    return bnOrZero(input.amountCryptoBaseUnit).gt(allowanceCryptoBaseUnit)
+  }, [allowanceDataQuery.data, chainId, input.amountCryptoBaseUnit, input.assetId])
+
+  const approvalFeesQuery = useEvmFees({
+    ...approvalFeesQueryInput,
+    enabled: Boolean(isApprovalRequired),
+    staleTime: 30_000,
+    // Ensures fees are refetched at an interval, including when the app is in the background
+    refetchIntervalInBackground: true,
+    // Yeah this is arbitrary but come on, Arb is cheap
```
Comment (gomesalexandre) [inline src/hooks/mutations/useApprove.ts:78]: This needs to be fixed before opening, this hook isn't Arb-specific

File: src/hooks/mutations/useApprove.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    refetchInterval: isApprovalRequired ? 15_000 : false,
+  })
+
+  const approveMutation = useMutation({
+    ...reactQueries.mutations.approve(maybeInputWithWallet),
+    onSuccess: (txId: string) => {
+      handleSuccess?.(txId)
+      if (!(input.assetId && input.spender && input.from)) return
```
Comment (gomesalexandre) [inline src/hooks/mutations/useApprove.ts:86]: Just to satisfy the type-system - we won't end up in here if these are undefined. No `skipToken` support for mutations, but these are asserted in the `approve` mutation.

File: src/pages/Lending/Pool/components/Repay/RepayInput.tsx [hunk 3]
```diff
@@ -131,27 +131,32 @@ export const RepayInput = ({
...
+    isApprovalRequired,
+    allowanceDataQuery: { isLoading: isAllowanceDataLoading },
+  } = useApprove({
     assetId: repaymentAsset?.assetId,
     spender: inboundAddressData?.router,
     from: userAddress,
+    amountCryptoBaseUnit:
+      confirmedQuote?.repaymentAmountCryptoPrecision && repaymentAsset
+        ? toBaseUnit(
+            // Add 5% buffer to the repayment allowance to avoid asset rates fluctuations ending up in more asset needed to repay
```
Comment (gomesalexandre) [inline src/pages/Lending/Pool/components/Repay/RepayInput.tsx:152]: Comment/logic just moved around but while at it, noticed we actually ditched it at some point for some reason, which broke things

File: src/pages/Lending/Pool/components/Repay/RepayInput.tsx [hunk 4]
```diff
@@ -464,15 +441,19 @@ export const RepayInput = ({
         )
         .lte(repaymentAssetAmountAvailableCryptoPrecision)
 
+    // Token - fees can either be network fees or Tx fees
```
Comment (gomesalexandre) [inline src/pages/Lending/Pool/components/Repay/RepayInput.tsx:444]:
```suggestion
    // Token - fees can either be network fees or approval fees
```

File: src/hooks/mutations/useApprove.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+type UseApproveProps = MaybeApproveInput & {
+  onSuccess?: (txId: string) => void
+}
+
+export const useApprove = ({ onSuccess: handleSuccess, ...input }: UseApproveProps) => {
+  const queryClient = useQueryClient()
+  const wallet = useWallet().state.wallet
+
+  const [approvalTxId, setApprovalTxId] = useState<string | null>(null)
```
Comment (gomesalexandre) [inline src/hooks/mutations/useApprove.ts:23]: `txHash` terminology

File: src/react-queries/queries/mutations.ts [hunk 6]
```diff
@@ -10,67 +10,82 @@ import {
...
+  amountCryptoBaseUnit: string
+  from: string
+  accountNumber: number
+}
+export type MaybeApproveInput = Partial<ApproveInput>
+
+type ApproveInputWithWallet = ApproveInput & { wallet: HDWallet }
+type MaybeApproveInputWithWallet = Partial<ApproveInputWithWallet>
+
+export const approve = async ({
```
Comment (gomesalexandre) [inline src/react-queries/queries/mutations.ts:25]: Now that this is moved around as a good ol' async fn, give me a better home

File: src/lib/utils/evm/approve.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
+import { fromAssetId } from '@shapeshiftoss/caip'
+import { CONTRACT_INTERACTION } from '@shapeshiftoss/chain-adapters'
+import { supportsETH } from '@shapeshiftoss/hdwallet-core'
+import type { ApproveInputWithWallet } from 'react-queries/queries/mutations'
+
+import { assertGetEvmChainAdapter, buildAndBroadcast, getApproveContractData, getFees } from '.'
+
+export const approve = async ({
```
Comment (gomesalexandre) [inline src/lib/utils/evm/approve.ts:8]: Not react-query specific anymore, just a good ol' async fn 🎉

File: src/lib/utils/evm/index.ts
Comment (gomesalexandre) [inline src/lib/utils/evm/index.ts:1]: figures out `lib/utils/evm` would be a perfect home for `approve()` but if we feel strong about not having an `evm` folder and keeping the previous `evm.ts` as a single source of truth for lower-level Tx-related utils, happy to revert and move `approve()` somewhere else

File: src/pages/Lending/Pool/components/Repay/RepayInput.tsx [hunk 8]
```diff
@@ -118,6 +110,14 @@ export const RepayInput = ({
...
   const collateralFeeAsset = useAppSelector(state => selectFeeAssetById(state, collateralAssetId))
 
+  const repaymentAccountNumberFilter = useMemo(
+    () => ({ accountId: repaymentAccountId ?? '' }),
+    [repaymentAccountId],
+  )
+  const repaymentAccountNumber = useAppSelector(state =>
+    selectAccountNumberByAccountId(state, repaymentAccountNumberFilter),
+  )
+
```
Comment (gomesalexandre) [inline src/pages/Lending/Pool/components/Repay/RepayInput.tsx:120]: Moved around

File: src/pages/Lending/Pool/components/Repay/RepayInput.tsx [hunk 9]
```diff
@@ -659,7 +640,7 @@ export const RepayInput = ({
         layout='inline'
         labelPostFix={collateralAssetSelectComponent}
       />
-      <Collapse in={isLendingQuoteCloseSuccess}>
+      <Collapse in={isLendingQuoteCloseSuccess} unmountOnExit>
```
Comment (gomesalexandre) [inline src/pages/Lending/Pool/components/Repay/RepayInput.tsx:643]: This was the culprit of rugged collapse with summary often not showing up

File: src/react-queries/queries/mutations.ts [hunk 10]
```diff
@@ -1,76 +1,39 @@
...
-    mutationKey: ['approve', { assetId, spender, amountCryptoBaseUnit, from }],
-    mutationFn: async (_: void) => {
+    wallet,
+  }: MaybeApproveInputWithWallet) => ({
+    mutationKey: ['approve', { accountNumber, assetId, spender, amountCryptoBaseUnit, from }],
+    mutationFn: (_: void) => {
       if (!assetId) throw new Error('assetId is required')
       if (!spender) throw new Error('spender is required')
-      if (!amountCryptoBaseUnit) throw new Error('amount is required')
+      if (amountCryptoBaseUnit === undefined) throw new Error('non-undefined amount is required')
```
Comment (gomesalexandre) [inline src/react-queries/queries/mutations.ts:31]: `approve('0')` is valid to revoke approvals

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Without looking too deeply at the code yet this does not seem to be working as expected.

On this branch I'm unable to approve a token, it always says "Repay". 
It also seems that the `jsonrpc` request to check if an approval is needed is never fired, not on initial asset selection or when changing the repay amount.

`develop`:

Screenshot 2024-06-14 at 12 18 06 PM

This PR:

Screenshot 2024-06-14 at 12 18 29 PM

File: src/lib/utils/evm/approve.ts (hunk 7 above)
Comment (0xApotheosis) [reply src/lib/utils/evm/approve.ts:8]: So good! Purity 🤝

**Review by 0xApotheosis — APPROVED** [review]
Ok, re-tested with a fresh cache, I think I was just in a bad local state before.

Approvals are happy and fees gracefully switch.

Screenshot 2024-06-17 at 10 52 02 AM
Screenshot 2024-06-17 at 10 52 10 AM

**Comment by 0xApotheosis** [comment]
Nothing blocking here, so I'm going to merge so we can keep this stack moving!

**Pull request merged**